### PR TITLE
complete config observer

### PR DESCRIPTION
**Pull request by juanvallejo**
cc @deads2k

### CODE REVIEW

File: pkg/operator/operator.go [hunk 1]
```diff
@@ -30,6 +30,7 @@ import (
 const (
 	kubeAPIServerNamespaceName = "openshift-kube-apiserver"
 	targetNamespaceName        = "openshift-controller-manager"
+	operatorNamespaceName      = "openshift-core-operators"
```
Comment (deads2k) [inline pkg/operator/operator.go]: missing an event handler on this namespace.

File: pkg/operator/observe_config.go [hunk 2]
```diff
@@ -56,16 +56,28 @@ func NewConfigObserver(
...
-// TODO(juanvallejo): update this to actually watch this operator's config
+// sync reacts to a change in controller manager images.
 func (c ConfigObserver) sync() error {
 	operatorConfig, err := c.operatorConfigClient.Get("instance", metav1.GetOptions{})
 	if err != nil {
 		return err
 	}
 	observedConfig := map[string]interface{}{}
 
+	controllerManagerImages, err := c.kubeClient.CoreV1().ConfigMaps(operatorNamespaceName).Get("openshift-controller-manager-images", metav1.GetOptions{})
```
Comment (deads2k) [inline pkg/operator/observe_config.go]: structure it with the observeConfig function describe here: https://github.com/openshift/cluster-kube-apiserver-operator/pull/74 with the additional `rest.Config` being passed.

File: pkg/operator/operator.go [hunk 3]
```diff
@@ -63,9 +65,11 @@ func NewOpenShiftControllerManagerOperator(
 	namespacedKubeInformers.Core().V1().ServiceAccounts().Informer().AddEventHandler(c.eventHandler())
 	namespacedKubeInformers.Core().V1().Services().Informer().AddEventHandler(c.eventHandler())
 	namespacedKubeInformers.Apps().V1().Deployments().Informer().AddEventHandler(c.eventHandler())
+	namespacedOperatorInformers.Core().V1().ConfigMaps().Informer().AddEventHandler(c.eventHandler())
```
Comment (juanvallejo) [inline pkg/operator/operator.go]: @deads2k not sure if I should also be watching deployments here?

File: pkg/operator/observe_config.go [hunk 4]
```diff
@@ -49,22 +53,30 @@ func NewConfigObserver(
...
-// sync reacts to a change in prereqs by finding information that is required to match another value in the cluster. This
-// must be information that is logically "owned" by another component.
-// TODO(juanvallejo): update this to actually watch this operator's config
+// sync reacts to a change in controller manager images.
 func (c ConfigObserver) sync() error {
+	var err error
+	observedConfig := map[string]interface{}{}
+
+	for _, observer := range c.observers {
+		observedConfig, err = observer(c.kubeClient, nil, observedConfig)
```
Comment (deads2k) [inline pkg/operator/observe_config.go]: plumb through a rest.Config to pass from the starter

File: pkg/operator/observe_config.go [hunk 5]
```diff
@@ -82,6 +94,25 @@ func (c ConfigObserver) sync() error {
...
 
+// observeControllerManagerImagesConfig observes image paths from openshift-controller-manager-images in order to determine which deployer and builder images to use
+func observeControllerManagerImagesConfig(kubeClient kubernetes.Interface, config *rest.Config, observedConfig map[string]interface{}) (map[string]interface{}, error) {
+	controllerManagerImages, err := kubeClient.CoreV1().ConfigMaps(operatorNamespaceName).Get("openshift-controller-manager-images", metav1.GetOptions{})
+	if err != nil && !errors.IsNotFound(err) {
+		return nil, err
+	}
+
+	if controllerManagerImages != nil {
+		if val, ok := controllerManagerImages.Data["builderImage"]; ok {
```
Comment (deads2k) [inline pkg/operator/observe_config.go]: check length, not ok.  make a TODO to reflect issues in status.

File: pkg/operator/operator.go [hunk 6]
```diff
@@ -47,6 +48,7 @@ type OpenShiftControllerManagerOperator struct {
 func NewOpenShiftControllerManagerOperator(
 	operatorConfigInformer operatorconfiginformerv1alpha1.OpenShiftControllerManagerOperatorConfigInformer,
 	namespacedKubeInformers informers.SharedInformerFactory,
```
Comment (deads2k) [inline pkg/operator/operator.go]: `kubeInformersFor<namespace>`

File: pkg/operator/operator.go [hunk 7]
```diff
@@ -47,6 +48,7 @@ type OpenShiftControllerManagerOperator struct {
...
+	namespacedOperatorInformers informers.SharedInformerFactory,
```
Comment (deads2k) [inline pkg/operator/operator.go]: `kubeInformersFor<namespace>`

File: pkg/operator/observe_config.go [hunk 8]
```diff
@@ -49,22 +53,30 @@ func NewConfigObserver(
 		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "ConfigObserver"),
 
 		rateLimiter: flowcontrol.NewTokenBucketRateLimiter(0.05 /*3 per minute*/, 4),
+		observers:   []observeConfigFunc{observeControllerManagerImagesConfig},
 	}
 
 	operatorConfigInformer.Informer().AddEventHandler(c.eventHandler())
```
Comment (deads2k) [inline pkg/operator/observe_config.go]: your event handler goes here.

File: pkg/operator/operator.go (hunk 3 above)
Comment (deads2k) [reply pkg/operator/operator.go]: > @deads2k not sure if I should also be watching deployments here?

I wasn't expecting a new trigger here.  This is the workload resource control loop. You aren't touching this one.

File: pkg/operator/starter.go [hunk 9]
```diff
@@ -42,10 +42,12 @@ func RunOperator(clientConfig *rest.Config, stopCh <-chan struct{}) error {
 
 	operatorConfigInformers := operatorclientinformers.NewSharedInformerFactory(operatorConfigClient, 10*time.Minute)
 	kubeInformersNamespaced := informers.NewFilteredSharedInformerFactory(kubeClient, 10*time.Minute, targetNamespaceName, nil)
+	coreInformersNamespaced := informers.NewFilteredSharedInformerFactory(kubeClient, 10*time.Minute, operatorNamespaceName, nil)
 
 	operator := NewOpenShiftControllerManagerOperator(
 		operatorConfigInformers.Openshiftcontrollermanager().V1alpha1().OpenShiftControllerManagerOperatorConfigs(),
 		kubeInformersNamespaced,
+		coreInformersNamespaced,
```
Comment (deads2k) [inline pkg/operator/starter.go]: wrong spot.  You need to wire it to the config observer control loop.

File: pkg/operator/observe_config.go [hunk 10]
```diff
@@ -82,6 +94,25 @@ func (c ConfigObserver) sync() error {
 	return nil
 }
 
+// observeControllerManagerImagesConfig observes image paths from openshift-controller-manager-images in order to determine which deployer and builder images to use
+func observeControllerManagerImagesConfig(kubeClient kubernetes.Interface, config *rest.Config, observedConfig map[string]interface{}) (map[string]interface{}, error) {
```
Comment (deads2k) [inline pkg/operator/observe_config.go]: test for this function please

**Comment by juanvallejo** [comment]
@deads2k thanks, added a test

File: pkg/operator/observe_config.go [hunk 11]
```diff
@@ -82,6 +100,26 @@ func (c ConfigObserver) sync() error {
...
+	controllerManagerImages, err := kubeClient.CoreV1().ConfigMaps(operatorNamespaceName).Get("openshift-controller-manager-images", metav1.GetOptions{})
+	if err != nil && !errors.IsNotFound(err) {
+		return nil, err
+	}
+
+	if controllerManagerImages != nil {
+		// TODO(juanvallejo): reflect any issues in operator status
+		if val, ok := controllerManagerImages.Data["builderImage"]; ok && len(val) > 0 {
```
Comment (deads2k) [inline pkg/operator/observe_config.go]: no  need to check ok

File: pkg/operator/observe_config_test.go [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+)
+
+func TestObserveClusterConfig(t *testing.T) {
+	tests := []struct {
+		name   string
+		cm     *v1.ConfigMap
+		expect map[string]interface{}
+	}{
+		{
+			name: "ensure valid configmap is observed and parsed",
```
Comment (deads2k) [inline pkg/operator/observe_config_test.go]: @sanchezl see table driven test here.

File: pkg/operator/observe_config_test.go [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+					"deployerImage": "quay.io/sample/origin-deployer:v4.0",
+				},
+			},
+			expect: map[string]interface{}{},
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			kubeClient := fake.NewSimpleClientset()
```
Comment (deads2k) [inline pkg/operator/observe_config_test.go]: initialize the clientset with the configmap directly. Using create adds an action you don't want.

File: pkg/operator/starter.go [hunk 14]
```diff
@@ -42,6 +42,7 @@ func RunOperator(clientConfig *rest.Config, stopCh <-chan struct{}) error {
 
 	operatorConfigInformers := operatorclientinformers.NewSharedInformerFactory(operatorConfigClient, 10*time.Minute)
 	kubeInformersNamespaced := informers.NewFilteredSharedInformerFactory(kubeClient, 10*time.Minute, targetNamespaceName, nil)
+	coreInformersNamespaced := informers.NewFilteredSharedInformerFactory(kubeClient, 10*time.Minute, operatorNamespaceName, nil)
```
Comment (deads2k) [inline pkg/operator/starter.go]: update to the construction sanchezl has in 74.

Make names consistent with `kubeInformersForOpenshiftControllerManager`

**Comment by juanvallejo** [comment]
@deads2k thanks, review comments addressed

File: pkg/operator/starter.go [hunk 15]
```diff
@@ -41,19 +41,22 @@ func RunOperator(clientConfig *rest.Config, stopCh <-chan struct{}) error {
 	ensureOperatorConfigExists(operatorConfigClient.OpenshiftcontrollermanagerV1alpha1(), "v3.11.0/openshift-controller-manager/operator-config.yaml")
 
 	operatorConfigInformers := operatorclientinformers.NewSharedInformerFactory(operatorConfigClient, 10*time.Minute)
-	kubeInformersNamespaced := informers.NewFilteredSharedInformerFactory(kubeClient, 10*time.Minute, targetNamespaceName, nil)
+	kubeInformersForOpenshiftOpenshiftControllerManagerNamespace := informers.NewSharedInformerFactoryWithOptions(kubeClient, 10*time.Minute, informers.WithNamespace(targetNamespaceName))
```
Comment (deads2k) [inline pkg/operator/starter.go]: OpenshiftOpenshfit typo

File: pkg/operator/starter.go [hunk 16]
```diff
@@ -41,19 +41,22 @@ func RunOperator(clientConfig *rest.Config, stopCh <-chan struct{}) error {
...
+	kubeInformersForOpenshiftCoreOperatorsNamespace := informers.NewSharedInformerFactoryWithOptions(kubeClient, 10*time.Minute, informers.WithNamespace(operatorNamespaceName))
```
Comment (deads2k) [inline pkg/operator/starter.go]: where is this informer started.

Comment (juanvallejo) [reply pkg/operator/starter.go]: Thanks, added a call to its Start method below

File: pkg/operator/starter.go (hunk 15 above)
Comment (juanvallejo) [reply pkg/operator/starter.go]: Thanks, fixed

**Pull request merged**